### PR TITLE
Fabo/100 node approval

### PR DESCRIPTION
**Pull request by faboweb**
> Thank you a lot for contributing to Cosmos Voyager!

### Issue

closes: #100
### Screenshots

Approval modal:
image

The second button will pick another node to connect to and show a new hash.

TODO
- [x] unit tests
- [x] e2e tests
- [x] design improvements
- [x] merge #572 

❤️ Thank you!

### CODE REVIEW

**Comment by nylira** [comment]
Let's align the buttons to the left, to match the text. And reduce the vertical spacing between the text field and the buttons by 50%. Otherwise it looks great!

File: app/src/renderer/components/common/NiModalLCDApproval.vue [hunk 1]
```diff
@@ -0,0 +1,135 @@
+<template lang="pug">
+.ni-modal-lcd-approval__wrapper
+  .ni-modal-lcd-approval
+    .ni-modal-lcd-approval__icon: i.material-icons sync problem
+    .ni-modal-lcd-approval__title Node approval required
+    .ni-modal-lcd-approval__body
+      | Voyager needs to approve the connected node. Please check the showing hash against a trusted source. Only approve if you are 100% sure the hash matches.
```
[the remark this answers is not in the document]
Comment (faboweb) [reply app/src/renderer/components/common/NiModalLCDApproval.vue]: Then we have failed our customers. ❤️

**Comment by nylira** [comment]
Looks great Fabo. Nice work!

**Comment by nylira** [comment]
@faboweb is this the same error you're getting with the e2e test?

```
pz@mbp ~/P/voyager> env BINARY_PATH=(which gaia) yarn test:e2e
yarn run v1.5.1
$ cross-env ANALYTICS=false yarn run pack && tape "test/e2e/*.js"
$ yarn run pack:main && yarn run pack:renderer
$ cross-env NODE_ENV=production webpack --colors --config webpack.main.config.js
Hash: 16df72dfe519378f81aa
Version: webpack 3.10.0
Time: 1216ms
  Asset    Size  Chunks                    Chunk Names
main.js  513 kB       0  [emitted]  [big]  main
 [103] ./config.js 1.17 kB {0} [built]
 [104] ./app/src/network.js 631 bytes {0} [built]
 [105] ./app/src/main/index.js 28.1 kB {0} [built]
 [108] (webpack)/buildin/module.js 517 bytes {0} [built]
 [222] ./node_modules/raven/lib/instrumentation ^\.\/.*$ 278 bytes {0} [built]
 [224] ./package.json 5.96 kB {0} [built]
 [225] ./app/src/main/menu.js 835 bytes {0} [built]
 [226] ./app/src/root.js 575 bytes {0} [built]
 [227] ./app/package.json 334 bytes {0} [built]
    + 219 hidden modules
$ cross-env NODE_ENV=production webpack --colors --config webpack.renderer.config.js
(node:49377) DeprecationWarning: loaderUtils.parseQuery() received a non-string value which can be problematic, see https://github.com/webpack/loader-utils/issues/56
parseQuery() will be replaced with getOptions() in the next major version of loader-utils.
Hash: a56ba39d666d4739ac01
Version: webpack 3.10.0
Time: 8956ms
                            Asset     Size  Chunks                    Chunk Names
  fonts/MaterialIcons-Regular.ttf   128 kB          [emitted]
    fonts/fontawesome-webfont.eot   166 kB          [emitted]
   fonts/fontawesome-webfont.woff    98 kB          [emitted]
    fonts/fontawesome-webfont.ttf   166 kB          [emitted]
     imgs/fontawesome-webfont.svg   444 kB          [emitted]  [big]
fonts/MaterialIcons-Regular.woff2  44.3 kB          [emitted]
 fonts/MaterialIcons-Regular.woff  57.6 kB          [emitted]
  fonts/fontawesome-webfont.woff2  77.2 kB          [emitted]
   imgs/cosmos-wordmark-black.svg  12.4 kB          [emitted]
   imgs/cosmos-wordmark-white.svg  12.4 kB          [emitted]
             imgs/cosmos-logo.png   125 kB          [emitted]
                      renderer.js  3.99 MB       0  [emitted]  [big]  renderer
                        vendor.js  5.78 kB       1  [emitted]         vendor
                       index.html   1.9 kB          [emitted]
 [176] ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./app/src/renderer/App.vue 2.14 kB {0} [built]
 [335] ./app/src/renderer/vuex/store.js 1.25 kB {0} [built]
 [452] ./app/src/renderer/main.js 5.16 kB {0} [built]
 [526] ./app/src/helpers/shrink-stacktrace.js 791 bytes {0} [built]
 [543] ./app/src/renderer/App.vue 1.48 kB {0} [built]
 [675] ./app/src/renderer/vuex/getters.js 1.82 kB {0} [built]
 [676] ./app/src/renderer/vuex/modules/index.js 824 bytes {0} [built]
 [740] ./node_modules/vue-loader/lib/template-compiler?{"id":"data-v-b508a830","hasScoped":false,"buble":{"transforms":{}}}!./node_modules/vue-loader/lib/template-compiler/preprocessor.js?engine=pug!./node_modules/vue-loader/lib/selector.js?type=template&index=0!./app/src/renderer/App.vue 660 bytes {0} [built]
 [741] ./app/src/renderer/routes.js 1.84 kB {0} [built]
 [742] ./app/src/renderer/components ^\.\/.*$ 4.19 kB {0} [built]
 [941] ./app/src/renderer/connectors/node.js 1.72 kB {0} [built]
 [948] ./app/src/renderer/connectors/lcdClient.js 7.98 kB {0} [built]
 [978] ./app/src/renderer/connectors/lcdClientMock.js 23.9 kB {0} [built]
 [979] ./app/src/renderer/connectors/rpcWrapper.js 1.54 kB {0} [built]
[1012] ./app/src/renderer/connectors/rpcWrapperMock.js 6.11 kB {0} [built]
    + 999 hidden modules
Child html-webpack-plugin for "index.html":
     1 asset
       [0] ./node_modules/html-webpack-plugin/lib/loader.js!./app/index.ejs 1.35 kB {0} [built]
       [2] (webpack)/buildin/module.js 517 bytes {0} [built]
        + 1 hidden module
TAP version 13
# sign in
using binary /Users/pz/go/bin/gaia
ui home: /var/folders/v3/yzp94ctx4dzdzcnvt1q4_7qm0000gn/T/z49gkzuy4ds
node home: /var/folders/v3/yzp94ctx4dzdzcnvt1q4_7qm0000gn/T/fnu8yd98lp/lcd
inited local node
started local node
ok 1 app is running
ok 2 app recovers from uninitialized gaia
restored test accounts
ok 3 app is running
restarting app
not ok 4 Error: ESOCKETTIMEDOUT
  ---
    operator: error
    expected: |-
      undefined
    actual: |-
      { [Error: ESOCKETTIMEDOUT] details: undefined, message: 'ESOCKETTIMEDOUT', type: 'RuntimeError', seleniumStack: { status: -1, type: 'ESOCKETTIMEDOUT', orgStatusMessage: 'ESOCKETTIMEDOUT', message: 'Couldn\'t connect to selenium server' } }
    at: bound (/Users/pz/Projects/voyager/node_modules/onetime/index.js:30:12)
    stack: |-
      Error: Couldn't connect to selenium server
          at execute(<Function>, "app", "quit", , "require") - api.js:269:21
  ...
# wallet
restarting app
```

File: app/src/main/index.js [hunk 2]
```diff
@@ -114,7 +117,8 @@ function createWindow() {
     webPreferences: { webSecurity: false }
   })
 
-  mainWindow.loadURL(winURL + "?node=" + nodeIP + "&lcd_port=" + LCD_PORT)
```
Comment (faboweb) [inline app/src/main/index.js]: nodeIP will be transmitted via IPC when available

File: app/src/main/index.js [hunk 3]
```diff
@@ -182,6 +185,8 @@ function startProcess(name, args, env) {
         err
       )
       handleCrash(err)
+
+      Raven.captureException(err)
```
Comment (faboweb) [inline app/src/main/index.js]: can be async as the app doesn't shutdown on errors anymore

**Comment by nylira** [comment]
Great work @faboweb!

File: app/src/main/index.js [hunk 4]
```diff
@@ -201,27 +206,35 @@ app.on("ready", () => createWindow())
...
+      "--home",
+      home,
+      "--node",
+      nodeIP
+      // '--trust-node'
+    ])
+    logProcess(child, join(home, "lcd.log"))
 
-  return child
+    // XXX why the hell stderr?!?!?!?
```
Comment (jolesbi) [inline app/src/main/index.js]: 🤣

File: app/src/main/index.js [hunk 5]
```diff
@@ -201,27 +206,35 @@ app.on("ready", () => createWindow())
...
+      if (data.includes("Serving on")) resolve(child)
+    })
+    child.on("exit", () => {
+      afterBooted(() => {
+        if (mainWindow) {
+          mainWindow.webContents.send(
+            "error",
+            Error("The Gaia REST-server (LCD) exited unplanned")
+          )
+        }
```
Comment (NodeGuy) [inline app/src/main/index.js]: Do you want to call `reject` here?

File: app/src/main/index.js [hunk 6]
```diff
@@ -201,27 +206,35 @@ app.on("ready", () => createWindow())
...
+    child.stderr.on("data", data => {
+      if (data.includes("Serving on")) resolve(child)
+    })
+    child.on("exit", () => {
+      afterBooted(() => {
+        if (mainWindow) {
+          mainWindow.webContents.send(
+            "error",
+            Error("The Gaia REST-server (LCD) exited unplanned")
+          )
```
Comment (NodeGuy) [inline app/src/main/index.js]: Do you want to call `reject` here?

Comment (faboweb) [reply app/src/main/index.js]: You're right. I will add that.

**Pull request merged**